File: pp_chia_doi.cpp
#include<iostream>
#include<cmath>
#include<iomanip>
using namespace std;
double abs(double x,double y){
    return x>y ? x-y:y-x;
}
double f_x(double x){
    return log(x)-1;
}
char sign(double a){
    if(a>0) return '+';
    else{
        return '-';
    }
    
}
void pp_chia_doi(double a,double b,double epxilon){
    double m;
    cout<<"STT        an        bn        xn        sign        denta_xn"<<endl;
    int cnt=1;
    while(abs(a,b)>epxilon){
        cout<<cnt<<"          "<<a<<"          "<<b<<"         "<<setprecision(15)<<((a+b)/2)<<"          "<<sign(f_x((a+b)/2))<<"          "<<(abs(a,b)/2)<<endl;
        m=(a+b)/2;
        if(f_x(m)*f_x(a)>0){
            a=m;
        }
        else{
            b=m;
        }
        cnt++;
    }
}

int main(){
    double a,b,epx;
    cout<<"NHAP VAO GIA TRI CUA a : ";cin>>a;
    cout<<"NHAP VAO GIA TRI CUA b : ";cin>>b;
    cout<<"NHAP VAO GIA TRI CUA epxilon : ";cin>>epx;
    pp_chia_doi(a,b,epx);
}

File: pp_chia_doi.py
import math

def sign(a):
    if(a>0):return "+"
    else:return "-"

def f_x(a):
    return math.log(a)-1

def pp_chia_doi(a,b,epxilon):
    cnt=0
    m=a
    n=b
    print("STT           a           b           xn            sign           denta_xn")
    while((n-m) > epxilon):
        c=(m+n)/2
        if(f_x(c) == 0):
            print("Nghiệm là ",c)
        else:
            print(cnt,"           ",m,"           ",n,"           ",c,"           ",sign(f_x(c)),"           ",(n-m)/2)
            if(f_x(c)*f_x(m)>0):
                m=c
            elif(f_x(c)*f_x(n)>0):
                n=c
        cnt+=1
# if __name__=="__main__":
a=float(input("NHAP a : "))
b=float(input("NHAP b : "))
epx=float(input("NHAP epxilon : "))
pp_chia_doi(a,b,epxilon=epx)

        

